File: scripts/process-wc26-logo.mjs
/**
 * Remove white background from official WC26 logo PNG.
 * Run: node scripts/process-wc26-logo.mjs [input.png] [output.png]
 */
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { createRequire } from "node:module";

const require = createRequire(import.meta.url);
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.join(__dirname, "..");

const input =
  process.argv[2] ??
  path.join(
    root,
    "assets/c__Users_LENOVO_AppData_Roaming_Cursor_User_workspaceStorage_9835331e563551428276856c879630b4_images_image-91643fc4-bd76-478e-8982-9d43132e1ad4.png",
  );
const output = process.argv[3] ?? path.join(root, "public/brand/wc26-logo.png");

const { PNG } = require("pngjs");

function lum(r, g, b) {
  return 0.299 * r + 0.587 * g + 0.114 * b;
}

function isBgWhite(r, g, b, a, t = 246) {
  return a > 0 && r >= t && g >= t && b >= t;
}

function loadPng(file) {
  return PNG.sync.read(fs.readFileSync(file));
}

function savePng(png, file) {
  fs.mkdirSync(path.dirname(file), { recursive: true });
  fs.writeFileSync(file, PNG.sync.write(png));
}

function idx(w, x, y) {
  return (w * y + x) << 2;
}

function floodBg(png) {
  const { width: w, height: h, data } = png;
  const seen = new Uint8Array(w * h);
  const q = [];

  for (let x = 0; x < w; x++) {
    q.push([x, 0], [x, h - 1]);
  }
  for (let y = 0; y < h; y++) {
    q.push([0, y], [w - 1, y]);
  }

  while (q.length) {
    const [x, y] = q.pop();
    if (x < 0 || y < 0 || x >= w || y >= h) continue;
    const i = y * w + x;
    if (seen[i]) continue;
    seen[i] = 1;
    const p = idx(w, x, y);
    const r = data[p];
    const g = data[p + 1];
    const b = data[p + 2];
    const a = data[p + 3];
    if (!isBgWhite(r, g, b, a)) continue;
    data[p + 3] = 0;
    q.push([x + 1, y], [x - 1, y], [x, y + 1], [x, y - 1]);
  }
}

/** Remove enclosed white holes inside the "26" numerals. */
function clearInteriorHoles(png) {
  const { width: w, height: h, data } = png;
  const visited = new Uint8Array(w * h);
  const components = [];

  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      const start = y * w + x;
      if (visited[start]) continue;
      const p = idx(w, x, y);
      const r = data[p];
      const g = data[p + 1];
      const b = data[p + 2];
      const a = data[p + 3];
      if (a === 0 || lum(r, g, b) < 235) {
        visited[start] = 1;
        continue;
      }

      const pixels = [];
      const q = [[x, y]];
      visited[start] = 1;
      let touchesEdge = false;

      while (q.length) {
        const [cx, cy] = q.pop();
        pixels.push([cx, cy]);
        if (cx === 0 || cy === 0 || cx === w - 1 || cy === h - 1) touchesEdge = true;
        for (const [nx, ny] of [
          [cx + 1, cy],
          [cx - 1, cy],
          [cx, cy + 1],
          [cx, cy - 1],
        ]) {
          if (nx < 0 || ny < 0 || nx >= w || ny >= h) continue;
          const ni = ny * w + nx;
          if (visited[ni]) continue;
          const np = idx(w, nx, ny);
          const nr = data[np];
          const ng = data[np + 1];
          const nb = data[np + 2];
          const na = data[np + 3];
          if (na === 0 || lum(nr, ng, nb) < 235) {
            visited[ni] = 1;
            continue;
          }
          visited[ni] = 1;
          q.push([nx, ny]);
        }
      }

      if (!touchesEdge && pixels.length > 120) {
        components.push(pixels);
      }
    }
  }

  for (const comp of components) {
    for (const [x, y] of comp) {
      const p = idx(w, x, y);
      data[p + 3] = 0;
    }
  }
}

function trimTransparent(png) {
  const { width: w, height: h, data } = png;
  let minX = w;
  let minY = h;
  let maxX = 0;
  let maxY = 0;

  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      if (data[idx(w, x, y) + 3] > 8) {
        minX = Math.min(minX, x);
        minY = Math.min(minY, y);
        maxX = Math.max(maxX, x);
        maxY = Math.max(maxY, y);
      }
    }
  }

  if (maxX < minX) return png;

  const pad = 4;
  minX = Math.max(0, minX - pad);
  minY = Math.max(0, minY - pad);
  maxX = Math.min(w - 1, maxX + pad);
  maxY = Math.min(h - 1, maxY + pad);
  const nw = maxX - minX + 1;
  const nh = maxY - minY + 1;
  const out = new PNG({ width: nw, height: nh });

  for (let y = 0; y < nh; y++) {
    for (let x = 0; x < nw; x++) {
      const sp = idx(w, x + minX, y + minY);
      const dp = idx(nw, x, y);
      out.data[dp] = data[sp];
      out.data[dp + 1] = data[sp + 1];
      out.data[dp + 2] = data[sp + 2];
      out.data[dp + 3] = data[sp + 3];
    }
  }
  return out;
}

if (!fs.existsSync(input)) {
  console.error("Input not found:", input);
  process.exit(1);
}

const png = loadPng(input);
floodBg(png);
clearInteriorHoles(png);
const trimmed = trimTransparent(png);
savePng(trimmed, output);
console.log("Wrote", output, `(${trimmed.width}x${trimmed.height})`);

File: scripts/process-wc26-logo.py
"""Remove white background from official WC26 logo. Run: python scripts/process-wc26-logo.py"""
from __future__ import annotations

import sys
from collections import deque
from pathlib import Path

from PIL import Image

ROOT = Path(__file__).resolve().parents[1]
DEFAULT_INPUT = ROOT / "assets" / "wc26-logo-source.jpg"
DEFAULT_OUTPUT = ROOT / "public" / "brand" / "wc26-logo.png"


def lum(r: int, g: int, b: int) -> float:
    return 0.299 * r + 0.587 * g + 0.114 * b


def is_bg_white(r: int, g: int, b: int, a: int, threshold: int = 246) -> bool:
    return a > 0 and r >= threshold and g >= threshold and b >= threshold


def flood_background(pixels, w: int, h: int) -> None:
    seen = bytearray(w * h)
    q: deque[tuple[int, int]] = deque()

    for x in range(w):
        q.append((x, 0))
        q.append((x, h - 1))
    for y in range(h):
        q.append((0, y))
        q.append((w - 1, y))

    while q:
        x, y = q.popleft()
        i = y * w + x
        if seen[i]:
            continue
        seen[i] = 1
        r, g, b, a = pixels[x, y]
        if not is_bg_white(r, g, b, a):
            continue
        pixels[x, y] = (255, 255, 255, 0)
        if x > 0:
            q.append((x - 1, y))
        if x + 1 < w:
            q.append((x + 1, y))
        if y > 0:
            q.append((x, y - 1))
        if y + 1 < h:
            q.append((x, y + 1))


def clear_interior_holes(pixels, w: int, h: int) -> None:
    visited = bytearray(w * h)

    for y in range(h):
        for x in range(w):
            start = y * w + x
            if visited[start]:
                continue
            r, g, b, a = pixels[x, y]
            if a == 0 or lum(r, g, b) < 235:
                visited[start] = 1
                continue

            comp: list[tuple[int, int]] = []
            q: deque[tuple[int, int]] = deque([(x, y)])
            visited[start] = 1
            touches_edge = False

            while q:
                cx, cy = q.popleft()
                comp.append((cx, cy))
                if cx == 0 or cy == 0 or cx == w - 1 or cy == h - 1:
                    touches_edge = True
                for nx, ny in ((cx + 1, cy), (cx - 1, cy), (cx, cy + 1), (cx, cy - 1)):
                    if nx < 0 or ny < 0 or nx >= w or ny >= h:
                        continue
                    ni = ny * w + nx
                    if visited[ni]:
                        continue
                    nr, ng, nb, na = pixels[nx, ny]
                    if na == 0 or lum(nr, ng, nb) < 235:
                        visited[ni] = 1
                        continue
                    visited[ni] = 1
                    q.append((nx, ny))

            if not touches_edge and len(comp) > 120:
                for cx, cy in comp:
                    pixels[cx, cy] = (255, 255, 255, 0)


def trim_transparent(img: Image.Image) -> Image.Image:
    bbox = img.getbbox()
    if not bbox:
        return img
    x0, y0, x1, y1 = bbox
    pad = 4
    x0 = max(0, x0 - pad)
    y0 = max(0, y0 - pad)
    x1 = min(img.width, x1 + pad)
    y1 = min(img.height, y1 + pad)
    return img.crop((x0, y0, x1, y1))


def remove_outer_fringe(pixels, w: int, h: int) -> None:
    """Strip JPEG white halos touching transparent pixels."""
    to_clear: list[tuple[int, int]] = []
    for y in range(h):
        for x in range(w):
            r, g, b, a = pixels[x, y]
            if a == 0 or lum(r, g, b) < 238:
                continue
            for nx, ny in ((x + 1, y), (x - 1, y), (x, y + 1), (x, y - 1)):
                if nx < 0 or ny < 0 or nx >= w or ny >= h:
                    to_clear.append((x, y))
                    break
                if pixels[nx, ny][3] == 0:
                    to_clear.append((x, y))
                    break
    for x, y in to_clear:
        pixels[x, y] = (255, 255, 255, 0)


def process(input_path: Path, output_path: Path) -> None:
    img = Image.open(input_path).convert("RGBA")
    pixels = img.load()
    w, h = img.size
    assert pixels is not None

    flood_background(pixels, w, h)
    clear_interior_holes(pixels, w, h)
    remove_outer_fringe(pixels, w, h)
    trimmed = trim_transparent(img)

    output_path.parent.mkdir(parents=True, exist_ok=True)
    trimmed.save(output_path, format="PNG", optimize=True)
    print(f"Wrote {output_path} ({trimmed.width}x{trimmed.height})")


if __name__ == "__main__":
    src = Path(sys.argv[1]) if len(sys.argv) > 1 else DEFAULT_INPUT
    dst = Path(sys.argv[2]) if len(sys.argv) > 2 else DEFAULT_OUTPUT
    if not src.exists():
        raise SystemExit(f"Input not found: {src}")
    process(src, dst)
